specs/deposits.md

```diff
@@ -0,0 +1,272 @@
+4. Person B wants to be sure that User A will pay for the computation.
+5. Person B wants to be paid for being middle man (for example cover gas costs and/or earn some money)
+
+Funder (User A) - Person that is funding the deposit and requesting job from the Spender service
+
+![funder](deposit_assets/funder300.webp)
+
+Spender (Person B) - Service based on yagna
+
+![spender](deposit_assets/spender300.webp)
+
+
+### Solution attempt allowance
+
+The simple solution is to use ERC20 allowance to person B's funds, but this solution has some drawbacks:
+Point 4 is not satisfied, because User A can revoke allowance at any time.
+Point 5 can be satisfied by taking funds beforehand.
+
+### Solution 
+
+Contract that will hold funds until specified time.
+
+![lock](deposit_assets/lock300.webp)
+
+We are using following interface:
+
+``` solidity
+struct Deposit {
+    address spender; //address that can spend the funds provided by customer
+    uint128 amount; //remaining funds locked
+    uint128 feeAmount; //fee amount locked for spender
+    uint64 validTo; //after this timestamp funds can be returned to customer
+}
+
+struct DepositView {
+    uint256 id;     //unique id
+    uint64 nonce;  //nonce unique for each funder
+    address funder; //address that can spend the funds provided by customer
+    address spender; //address that can spend the funds provided by customer
+    uint128 amount; //remaining funds locked
+    uint128 feeAmount; //fee amount locked for spender
+    uint64 validTo; //after this timestamp funds can be returned to customer
+}
+
+interface ILockPayment {
+    // Spender can close deposit anytime claiming fee and returning rest of funds to Funder
+    function closeDeposit(uint256 id) external;
+    // Funder can terminate deposit after validTo date elapses
+    function terminateDeposit(uint64 nonce) external;
+    function depositSingleTransfer(uint256 id, address addr, uint128 amount) external;
+    function depositTransfer(uint256 id, bytes32[] calldata payments) external;
+    function depositSingleTransferAndClose(uint256 id, address addr, uint128 amount) external;
+    function depositTransferAndClose(uint256 id, bytes32[] calldata payments) external;
+    function getMyDeposit(uint64 nonce) external view returns (DepositView memory);
+    function getDeposit(uint256 id) external view returns (DepositView memory);
+    function getDepositByNonce(uint64 nonce, address funder) external view returns (DepositView memory);
+}
+
+```
+
+Additional events emitted by the contract (for easier web3 integration). There is no need for
+extra options emitted by the contract, because additional information can be extracted using getDeposit(id) function.
+
+``` solidity
+    event DepositCreated(uint256 id, address spender);
+    event DepositExtended(uint256 id, address spender);
+    event DepositClosed(uint256 id, address spender);
+    event DepositTerminated(uint256 id, address spender);
+```
```

Comment:
We should explain that these events serve a purpose of replacing lack of list deposits function on contract